client/log_power.py

```diff
@@ -30,6 +30,59 @@ def read_measurement():
     sp.check_returncode()
     return sp.stdout
 
+def read_rpm(set_threshold_value, set_sampling_rate):
+
+    #should be at 0V (set to 2048 which should be half of the ADC range)
+    set_th_rising = subprocess.run([f"{DIRECTORY}/micromini-tool", "set-ain-rising-threshold", set_threshold_value], capture_output=False)
+    set_th_rising.check_returncode()
+    set_th_falling = subprocess.run([f"{DIRECTORY}/micromini-tool", "set-ain-falling-threshold", set_threshold_value], capture_output=False)
+    set_th_falling.check_returncode()
+    
+    #should be rougly at 170HZ such that we can measure between 5 and 1000 rpm with 400 samples (set to 98 which should come out to this frequency)
+    set_sr = subprocess.run([f"{DIRECTORY}/micromini-tool", "set-ain-rate", set_sampling_rate], capture_output=False)
+    set_sr.check_returncode()
+
+    sp_rising = subprocess.run([f"{DIRECTORY}/micromini-tool", "get-ain-nrising"], capture_output=True)
+    sp_rising.check_returncode()
+    sp_falling = subprocess.run([f"{DIRECTORY}/micromini-tool", "get-ain-nfalling"], capture_output=True)
+    sp_falling.check_returncode()
+    sp_samples = subprocess.run([f"{DIRECTORY}/micromini-tool", "get-ain-num-samples"], capture_output=True)
+    sp_samples.check_returncode()
+    sp_sampling_rate = subprocess.run([f"{DIRECTORY}/micromini-tool", "get-ain-rate"], capture_output=True)
+    sp_sampling_rate.check_returncode()
+
+    rising_crossings = float(sp_rising.stdout.decode("utf-8").strip().split(":")[1].strip())
+    falling_crossings = float(sp_falling.stdout.decode("utf-8").strip().split(":")[1].strip())
+    num_samples = float(sp_samples.stdout.decode("utf-8").strip())
+
+    #"Get AIN rate configuration (bits 0-2: ADC prescaler, bits 3-7 SAMPLELEN)"
+    raw = int(sp_sampling_rate.stdout.decode("utf-8").strip().split(":")[1].strip())
+
+    prescaler_bits = raw & 0b00000111
+    samplelen_bits = (raw >> 3) & 0b00011111
+
+    prescaler_dividers = [2, 4, 8, 16, 32, 64, 128, 256]
+    divider = prescaler_dividers[prescaler_bits]
+
+    adc_clock_hz = 32768
+    total_cycles = samplelen_bits + 13
```

Comment:
I assume you got this from reading the SAMD21 datasheet :) I can't remember the details here...

Reply:
Yes exactly. I'm not 100% sure about the '13' as it is not explicitly written in the data sheet, but I also found some code that used this in a similar way.